_posts/2022-08-30-solving-problems-with-extensions-2.adoc

```diff
@@ -0,0 +1,193 @@
+
+ArC, our CDI implementation, is one of the components that consumes the Jandex indexes.
+
+Interestingly though, ArC does not consume the Jandex index as is:
+
+image::annotations-transformers.svg[]
+
+Annotations transformers can add, remove, update existing annotations before consumption by ArC.
+These are used by several features in Quarkus, for instance Hibernate Validator interceptor support.
+
+This behavior is of great interest to us:
+we could hide the annotations from ArC using an annotation transformer
+while keeping them available for Airline to consume them via reflection.
+
+Let's create our annotations transformer:
+
+[source,java]
+----
+public class HideAirlineInjectAnnotationsTransformer implements AnnotationsTransformer { <1>
+
+    private final IndexView index;
+
+    HideAirlineInjectAnnotationsTransformer(IndexView index) { <2>
+        this.index = index;
+    }
+
+    @Override
+    public boolean appliesTo(Kind kind) {
+        return Kind.FIELD == kind; <3>
+    }
+
+    @Override
+    public void transform(TransformationContext transformationContext) {
+        FieldInfo fieldInfo = transformationContext.getTarget().asField();
+
+        if (!fieldInfo.hasAnnotation(DotNames.INJECT)) { <4>
+            return;
+        }
+
+        if (fieldInfo.hasAnnotation(ARGUMENTS) ||
+                fieldInfo.hasAnnotation(OPTION) ||
+                GLOBAL_METADATA.equals(fieldInfo.type().name()) || <5>
+                COMMAND_GROUP_METADATA.equals(fieldInfo.type().name()) ||
+                COMMAND_METADATA.equals(fieldInfo.type().name()) ||
+                isComposition(fieldInfo)) { <6>
+            transformationContext.transform().remove(ai -> DotNames.INJECT.equals(ai.name())).done(); <7>
+        }
+    }
+
+    private boolean isComposition(FieldInfo fieldInfo) { <8>
+        Type fieldType = fieldInfo.type();
+
+        if (fieldType.kind() != Type.Kind.CLASS) {
+            return false;
+        }
+
+        ClassInfo fieldClass = index.getClassByName(fieldType.asClassType().name());
+
+        if (fieldClass == null) {
+            return false;
+        }
+
+        Set<DotName> fieldClassAnnotations = fieldClass.annotationsMap().keySet();
+
+        return fieldClassAnnotations.contains(ARGUMENTS) || fieldClassAnnotations.contains(OPTION);
+    }
+}
+----
+<1> Our class implements `AnnotationsTransformer`.
+<2> We inject the Jandex index in our transformer as we will need it to detect composition.
+<3> We are only interested in fields so let's apply our transformer to fields only.
+<4> If the field is not annotated with `@Inject`, it is of no interest to us.
+<5> If the field type is `GlobalMetadata`, `GroupMetadata` or `CommandMetadata`, we know it is the responsibility of Airline to inject it.
+<6> We are also detecting composition.
+<7> We remove the `@Inject` annotation in this case. Make sure you don't forget to finalize the transformation with `.done()`.
```

Comment:
I was surprised at first that this wouldn't break Airline injection, but of course that's because we don't actually remove the annotation from the class, we just remove it from the index used by Arc. Maybe that should be stressed out?

Reply:
Yeah, it was in the schema but I added some text to make it extra clear.

Reply:
Thanks for the review!